File: src/libnxcl/alarms.cpp
/* 
** NetXMS - Network Management System
** Client Library
** Copyright (C) 2003-2010 Victor Kirhenshtein
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU Lesser General Public License as published by
** the Free Software Foundation; either version 3 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU Lesser General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
** File: alarms.cpp
**
**/

#include "libnxcl.h"


//
// Fill alarm record from message
//

static void AlarmFromMsg(CSCPMessage *pMsg, NXC_ALARM *pAlarm)
{
   pAlarm->dwAckByUser = pMsg->GetVariableLong(VID_ACK_BY_USER);
   pAlarm->dwTermByUser = pMsg->GetVariableLong(VID_TERMINATED_BY_USER);
   pAlarm->qwSourceEventId = pMsg->GetVariableInt64(VID_EVENT_ID);
   pAlarm->dwSourceEventCode = pMsg->GetVariableLong(VID_EVENT_CODE);
   pAlarm->dwSourceObject = pMsg->GetVariableLong(VID_OBJECT_ID);
   pAlarm->dwCreationTime = pMsg->GetVariableLong(VID_CREATION_TIME);
   pAlarm->dwLastChangeTime = pMsg->GetVariableLong(VID_LAST_CHANGE_TIME);
   pMsg->GetVariableStr(VID_ALARM_KEY, pAlarm->szKey, MAX_DB_STRING);
   pMsg->GetVariableStr(VID_ALARM_MESSAGE, pAlarm->szMessage, MAX_DB_STRING);
   pAlarm->nState = (BYTE)pMsg->GetVariableShort(VID_STATE);
   pAlarm->nCurrentSeverity = (BYTE)pMsg->GetVariableShort(VID_CURRENT_SEVERITY);
   pAlarm->nOriginalSeverity = (BYTE)pMsg->GetVariableShort(VID_ORIGINAL_SEVERITY);
   pAlarm->dwRepeatCount = pMsg->GetVariableLong(VID_REPEAT_COUNT);
   pAlarm->nHelpDeskState = (BYTE)pMsg->GetVariableShort(VID_HELPDESK_STATE);
   pMsg->GetVariableStr(VID_HELPDESK_REF, pAlarm->szHelpDeskRef, MAX_HELPDESK_REF_LEN);
	pAlarm->dwTimeout = pMsg->GetVariableLong(VID_ALARM_TIMEOUT);
	pAlarm->dwTimeoutEvent = pMsg->GetVariableLong(VID_ALARM_TIMEOUT_EVENT);
   pAlarm->pUserData = NULL;
}


//
// Process CMD_ALARM_UPDATE message
//

void ProcessAlarmUpdate(NXCL_Session *pSession, CSCPMessage *pMsg)
{
   NXC_ALARM alarm;
   DWORD dwCode;

   dwCode = pMsg->GetVariableLong(VID_NOTIFICATION_CODE);
   alarm.dwAlarmId = pMsg->GetVariableLong(VID_ALARM_ID);
   AlarmFromMsg(pMsg, &alarm);
   pSession->callEventHandler(NXC_EVENT_NOTIFICATION, dwCode, &alarm);
}


//
// Load all alarms from server
//

DWORD LIBNXCL_EXPORTABLE NXCLoadAllAlarms(NXC_SESSION hSession, BOOL bIncludeAck,
                                          DWORD *pdwNumAlarms, NXC_ALARM **ppAlarmList)
{
   CSCPMessage msg, *pResponse;
   DWORD dwRqId, dwRetCode = RCC_SUCCESS, dwNumAlarms = 0, dwAlarmId = 0;
   NXC_ALARM *pList = NULL;

   dwRqId = ((NXCL_Session *)hSession)->CreateRqId();

   msg.SetCode(CMD_GET_ALL_ALARMS);
   msg.SetId(dwRqId);
   msg.SetVariable(VID_IS_ACK, (WORD)bIncludeAck);
   ((NXCL_Session *)hSession)->SendMsg(&msg);

   do
   {
      pResponse = ((NXCL_Session *)hSession)->WaitForMessage(CMD_ALARM_DATA, dwRqId);
      if (pResponse != NULL)
      {
         dwAlarmId = pResponse->GetVariableLong(VID_ALARM_ID);
         if (dwAlarmId != 0)  // 0 is end of list indicator
         {
            pList = (NXC_ALARM *)realloc(pList, sizeof(NXC_ALARM) * (dwNumAlarms + 1));
            pList[dwNumAlarms].dwAlarmId = dwAlarmId;
            AlarmFromMsg(pResponse, &pList[dwNumAlarms]);
            dwNumAlarms++;
         }
         delete pResponse;
      }
      else
      {
         dwRetCode = RCC_TIMEOUT;
         dwAlarmId = 0;
      }
   }
   while(dwAlarmId != 0);

   // Destroy results on failure or save on success
   if (dwRetCode == RCC_SUCCESS)
   {
      *ppAlarmList = pList;
      *pdwNumAlarms = dwNumAlarms;
   }
   else
   {
      safe_free(pList);
      *ppAlarmList = NULL;
      *pdwNumAlarms = 0;
   }

   return dwRetCode;
}


//
// Acknowledge alarm by ID
//

DWORD LIBNXCL_EXPORTABLE NXCAcknowledgeAlarm(NXC_SESSION hSession, DWORD dwAlarmId)
{
   CSCPMessage msg;
   DWORD dwRqId;

   dwRqId = ((NXCL_Session *)hSession)->CreateRqId();

   msg.SetCode(CMD_ACK_ALARM);
   msg.SetId(dwRqId);
   msg.SetVariable(VID_ALARM_ID, dwAlarmId);
   ((NXCL_Session *)hSession)->SendMsg(&msg);

   return ((NXCL_Session *)hSession)->WaitForRCC(dwRqId);
}


//
// Terminate alarm by ID
//

DWORD LIBNXCL_EXPORTABLE NXCTerminateAlarm(NXC_SESSION hSession, DWORD dwAlarmId)
{
   CSCPMessage msg;
   DWORD dwRqId;

   dwRqId = ((NXCL_Session *)hSession)->CreateRqId();

   msg.SetCode(CMD_TERMINATE_ALARM);
   msg.SetId(dwRqId);
   msg.SetVariable(VID_ALARM_ID, dwAlarmId);
   ((NXCL_Session *)hSession)->SendMsg(&msg);

   return ((NXCL_Session *)hSession)->WaitForRCC(dwRqId);
}


//
// Delete alarm by ID
//

DWORD LIBNXCL_EXPORTABLE NXCDeleteAlarm(NXC_SESSION hSession, DWORD dwAlarmId)
{
   CSCPMessage msg;
   DWORD dwRqId;

   dwRqId = ((NXCL_Session *)hSession)->CreateRqId();

   msg.SetCode(CMD_DELETE_ALARM);
   msg.SetId(dwRqId);
   msg.SetVariable(VID_ALARM_ID, dwAlarmId);
   ((NXCL_Session *)hSession)->SendMsg(&msg);

   return ((NXCL_Session *)hSession)->WaitForRCC(dwRqId);
}


//
// Set helpdesk state to "Open"
//

DWORD LIBNXCL_EXPORTABLE NXCOpenAlarm(NXC_SESSION hSession, DWORD dwAlarmId, TCHAR *pszHelpdeskRef)
{
   CSCPMessage msg;
   DWORD dwRqId;

   dwRqId = ((NXCL_Session *)hSession)->CreateRqId();

   msg.SetCode(CMD_SET_ALARM_HD_STATE);
   msg.SetId(dwRqId);
   msg.SetVariable(VID_ALARM_ID, dwAlarmId);
   msg.SetVariable(VID_HELPDESK_STATE, (WORD)ALARM_HELPDESK_OPEN);
	msg.SetVariable(VID_HELPDESK_REF, CHECK_NULL_EX(pszHelpdeskRef));
   ((NXCL_Session *)hSession)->SendMsg(&msg);

   return ((NXCL_Session *)hSession)->WaitForRCC(dwRqId);
}


//
// Set helpdesk state to "Closed"
//

DWORD LIBNXCL_EXPORTABLE NXCCloseAlarm(NXC_SESSION hSession, DWORD dwAlarmId)
{
   CSCPMessage msg;
   DWORD dwRqId;

   dwRqId = ((NXCL_Session *)hSession)->CreateRqId();

   msg.SetCode(CMD_SET_ALARM_HD_STATE);
   msg.SetId(dwRqId);
   msg.SetVariable(VID_ALARM_ID, dwAlarmId);
   msg.SetVariable(VID_HELPDESK_STATE, (WORD)ALARM_HELPDESK_CLOSED);
   ((NXCL_Session *)hSession)->SendMsg(&msg);

   return ((NXCL_Session *)hSession)->WaitForRCC(dwRqId);
}


//
// Format text from alarm data
// Valid format specifiers are following:
//		%a Primary IP address of source object
//		%A Primary host name of source object
//    %c Repeat count
//    %e Event code
//    %E Event name
//    %h Helpdesk state as number
//    %H Helpdesk state as text
//    %i Source object identifier
//    %I Alarm identifier
//    %m Message text
//    %n Source object name
//    %s Severity as number
//    %S Severity as text
//    %% Percent sign
//

TCHAR LIBNXCL_EXPORTABLE *NXCFormatAlarmText(NXC_SESSION session, NXC_ALARM *alarm, TCHAR *format)
{
	static const TCHAR *helpdeskState[] = { _T("IGNORED"), _T("OPEN"), _T("CLOSED") };
	static const TCHAR *severityText[] = { _T("NORMAL"), _T("WARNING"), _T("MINOR"), _T("MAJOR"), _T("CRITICAL") };

	NXC_OBJECT *object = NXCFindObjectById(session, alarm->dwSourceObject);
	if (object == NULL)
	{
		NXCSyncSingleObject(session, alarm->dwSourceObject);
		object = NXCFindObjectById(session, alarm->dwSourceObject);
	}

	String out;
	TCHAR *prev, *curr, ipAddr[32];
	for(prev = format; *prev != 0; prev = curr)
	{
		curr = _tcschr(prev, _T('%'));
		if (curr == NULL)
		{
			out += prev;
			break;
		}
		out.addString(prev, (DWORD)(curr - prev));
		curr++;
		switch(*curr)
		{
			case '%':
				out += _T("%");
				break;
			case 'a':
				out += (object != NULL) ? IpToStr(object->dwIpAddr, ipAddr) : _T("<unknown>");
				break;
			case 'A':
				out += (object != NULL) ? object->node.szPrimaryName : _T("<unknown>");
				break;
			case 'c':
				out.addFormattedString(_T("%u"), (unsigned int)alarm->dwRepeatCount);
				break;
			case 'e':
				out.addFormattedString(_T("%u"), (unsigned int)alarm->dwSourceEventCode);
				break;
			case 'E':
				out += NXCGetEventName(session, alarm->dwSourceEventCode);
				break;
			case 'h':
				out.addFormattedString(_T("%d"), (int)alarm->nHelpDeskState);
				break;
			case 'H':
				out += helpdeskState[alarm->nHelpDeskState];
				break;
			case 'i':
				out.addFormattedString(_T("%u"), (unsigned int)alarm->dwSourceObject);
				break;
			case 'I':
				out.addFormattedString(_T("%u"), (unsigned int)alarm->dwAlarmId);
				break;
			case 'm':
				out += alarm->szMessage;
				break;
			case 'n':
				out += (object != NULL) ? object->szName : _T("<unknown>");
				break;
			case 's':
				out.addFormattedString(_T("%d"), (int)alarm->nCurrentSeverity);
				break;
			case 'S':
				out += severityText[alarm->nCurrentSeverity];
				break;
			case 0:
				curr--;
				break;
			default:
				break;
		}
		curr++;
	}
	return _tcsdup((const TCHAR *)out);
}

File: src/server/nxapisrv/java/src/main/java/org/netxms/websvc/handlers/Alarms.java
/**
 * NetXMS - open source network management system
 * Copyright (C) 2003-2017 Raden Solutions
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package org.netxms.websvc.handlers;

import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.UUID;
import org.json.JSONArray;
import org.json.JSONObject;
import org.netxms.client.NXCException;
import org.netxms.client.NXCSession;
import org.netxms.client.constants.RCC;
import org.netxms.client.events.Alarm;
import org.netxms.client.objects.AbstractObject;
import org.netxms.websvc.json.ResponseContainer;
import org.restlet.data.MediaType;
import org.restlet.ext.json.JsonRepresentation;
import org.restlet.representation.Representation;
import org.restlet.representation.StringRepresentation;
import org.restlet.resource.Post;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Handler for alarm management
 */
public class Alarms extends AbstractHandler
{
   private static final int TERMINATE = 0;
   private static final int ACKNOWLEDGE = 1;
   private static final int STICKY_ACKNOWLEDGE = 2;
   private static final int RESOLVE = 3;
   
   private Logger log = LoggerFactory.getLogger(Alarms.class);
   
   /* (non-Javadoc)
    * @see org.netxms.websvc.handlers.AbstractHandler#getCollection(org.json.JSONObject)
    */
   @Override
   protected Object getCollection(Map<String, String> query) throws Exception
   {
      NXCSession session = getSession();
      Collection<Alarm> alarms = session.getAlarms().values();
      String queryGuid = query.get("objectGuid");
      if(queryGuid != null)
      {         
         UUID objectGuid = UUID.fromString(queryGuid);
         if (!session.isObjectsSynchronized())
            session.syncObjects();
         
         AbstractObject object = session.findObjectByGUID(objectGuid);
         if (object == null)
            throw new NXCException(RCC.INVALID_OBJECT_ID);
         
         Iterator<Alarm> iterator =  alarms.iterator();
         while(iterator.hasNext())
         {
            Alarm alarm = iterator.next();
            if(alarm.getSourceObjectId() != object.getObjectId())
            {
               iterator.remove();
            }
         }
      }

      return new ResponseContainer("alarms", alarms);
   }
   
   @Post
   public Representation onPost(Representation entity) throws Exception
   {
      if (entity == null || !attachToSession())
      {
         return new StringRepresentation(createErrorResponse(RCC.ACCESS_DENIED).toString(), MediaType.APPLICATION_JSON);
      }
      
      NXCSession session = getSession();
      HashMap<Long, Alarm> alarms = session.getAlarms();
      
      JSONObject data = new JsonRepresentation(entity).getJsonObject();
      log.debug(data.toString());
      
      if (data.has("alarmList"))
      {
         JSONArray alarmList = data.getJSONArray("alarmList");
         
         for(int i = 0; i < alarmList.length(); i++)
         {
            JSONObject alarmData = alarmList.getJSONObject(i);
            Alarm a = alarms.get(alarmData.getLong("alarmId"));
            if (a != null)
            {
               switch (alarmData.getInt("action"))
               {
                  case TERMINATE:
                     session.terminateAlarm(a.getId());
                     break;
                  case ACKNOWLEDGE:
                     session.acknowledgeAlarm(a.getId());
                     break;
                  case STICKY_ACKNOWLEDGE:
                     if (alarmData.has("timeout"))
                        session.acknowledgeAlarm(a.getId(), true, alarmData.getInt("timeout"));
                     break;
                  case RESOLVE:
                     session.resolveAlarm(a.getId());
                     break;
               }
            }
         }
      }
      else
         return new StringRepresentation(createErrorResponse(RCC.INTERNAL_ERROR).toString(), MediaType.APPLICATION_JSON);
      
      return new StringRepresentation(createErrorResponse(RCC.SUCCESS).toString(), MediaType.APPLICATION_JSON);
   }
}
